[MIG] sale_order_price_recalculation: Migration to 12.0

rven opened the pull request.

pedrobaeza committed:
[ADD] sale_order_recalculate_prices: Module to allow to recalculate on demand sale order line prices

==================================================
Recalculation of sale order lines prices on demand
==================================================

This module adds a button on sale orders below pricelist that recalculates the
prices of the order lines that contain a product in them.

It is launched manually as a button to get the user decide if he/she wants to
recalculate prices when pricelist is changed.

Usage
=====

Inside a sale order, you can click in any moment a button called
"(Recalculate prices)", that is next to the pricelist selection, to launch
a recalculation of all the prices of the lines, losing previous custom prices.

Known issues / Roadmap
======================

* In a sale order with lot of lines, the recalculation may slow down, because
  sale general data (amount untaxed, amount taxed...) are recalculated for
  each line.

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[IMP] sale_order_price_recalculation

* add choice to recompute price only
  This commit add the capability to reset sale order line
  without renaming sale order lines
* upgrade translation
* Split functionnalities reset prices/reset desciptions
* Move button under sale order lines

cubells committed:
sale_order_price_recalculation module

chienandalu committed:
[MIG] sale_order_price_recalculation: Migration to 10.0

Commit:
[MIG] sale_order_price_recalculation: Migration to 11.0

oca-travis committed:
[UPD] Update sale_order_price_recalculation.pot

fmdl committed:
[FIX] position of button is not good with note field

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

pedrobaeza commented:
There are other files touched in this PR. I think you haven't performed correctly the procedure for extracting commits.

pedrobaeza commented on sale_order_price_recalculation/views/sale_order_view.xml:
```diff
@@ -0,0 +1,26 @@
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form"/>
+        <field name="arch" type="xml">
+            <field name="order_line" position="after">
```
Please use here:
```suggestion
            <field name="note" position="after">
```
the same as in:

https://github.com/OCA/commission/blob/64dc262a10a408c10e54a1cf3e36b74289dc7bc4/sale_commission/views/sale_order_view.xml#L33

as I detected this in that module that moves the total and taxes summary very down out of sight.

rven commented:
> There are other files touched in this PR. I think you haven't performed correctly the procedure for extracting commits.

I squashed some translation commits... maybe they caused the files to be modified

rven commented:
@pedrobaeza I don't see anything strange in my commit:
https://github.com/OCA/sale-workflow/pull/773/commits/845970773e12ed921bc049065d23386976c9dd0f
Here only the files I touched are shown.

pedrobaeza commented:
It's not on your migration commit, but on previous one.

![seleccion_020](https://user-images.githubusercontent.com/7165771/50403268-7c689780-079d-11e9-8f7f-61e7fc6d26f6.png)

Maybe an outdated rebase?

rven commented:
@pedrobaeza does this look better now?

pedrobaeza commented:
It seems so, thanks

pedrobaeza commented on sale_order_price_recalculation/models/sale_order.py:
```diff
@@ -1,8 +1,3 @@
-# Copyright 2014 Carlos Sánchez Cifuentes <csanchez@grupovermon.com>
```
Why removing here copyright? You should do it only on `__init__.py` files.

rven replied:
Okay, I misunderstood another PR where I needed to remove these Copyrights, but it was probably only the __init__.py file

pedrobaeza commented on sale_order_price_recalculation/views/sale_order_view.xml:
```diff
@@ -6,17 +6,17 @@
                                 string="Recalculate prices"
                                 attrs="{'invisible': [('state', 'not in', ['draft', 'sent'])]}"
-                                type="object" colspan="4" class="btn-default"/>
+                                type="object" colspan="4" class="ml-2 btn btn-primary"/>
```
Don't put it as primary button.

pedrobaeza commented on sale_order_price_recalculation/views/sale_order_view.xml:
```diff
@@ -6,17 +6,17 @@
                                 string="Reset descriptions"
                                 attrs="{'invisible': [('state', 'not in', ['draft', 'sent'])]}"
-                                type="object" colspan="4" class="btn-default"/>
+                                type="object" colspan="4" class="ml-2 btn btn-primary"/>
```
Not primary

rven committed:
[MIG] sale_order_price_recalculation: Migration to 12.0

pedrobaeza committed:
[IMP] sale_order_price_recalculation: Make tests deterministic

On certain set of installed modules, standard demo pricelist can be altered for giving
something different than the list price, making these tests to fail. We avoid it
creating our own pricelist with the fixed data that we want.

pedrobaeza commented:
Please include https://github.com/OCA/sale-workflow/commit/a3f7b91a3a744c909d04cad19394a7180ff6d603 that I have found that can be a problem in integration tests.

NL66278 commented on sale_order_price_recalculation/README.rst:
```diff
@@ -0,0 +1,113 @@
+
+* In a sale order with lot of lines, the recalculation may slow down, because
+  sale general data (amount untaxed, amount taxed...) are recalculated for
+  each line.
```
You might have a look at the two methods here to boost performance:
- disable prefetch of fields
- delay recompute until all changes that affect the computation have been done, that is in this case, until after all lines have been updated: https://github.com/OCA/contract/blob/f9c3e98a274610bb21be1b81f81c940b0ed131d1/contract/models/account_analytic_account.py#L325

pedrobaeza replied:
Good ideas!

rven replied:
`    
    @api.multi
    def recalculate_prices(self):
        _self = self.with_context(prefetch_fields=False)
        with _self.env.norecompute():
            for line in _self.mapped('order_line'):
                dict = line._convert_to_write(line.read()[0])
                if 'product_tmpl_id' in line._fields:
                    dict['product_tmpl_id'] = line.product_tmpl_id
                line2 = _self.env['sale.order.line'].new(dict)
                # we make this to isolate changed values:
                line2.product_uom_change()
                line.price_unit = line2.price_unit
        _self.recompute()
        return True

    @api.multi
    def recalculate_names(self):
        _self = self.with_context(prefetch_fields=False)
        with _self.env.norecompute():
            for line in _self.mapped('order_line').filtered('product_id'):
                # we make this to isolate changed values:
                line2 = _self.env['sale.order.line'].new({
                    'product_id': line.product_id,
                })
                line2.product_id_change()
                line.name = line2.name
        _self.recompute()
        return True`

Like this?
Do I have to recompute the order lines also?

NL66278 replied:
I think this should be enough, as you already update the move lines with the information you want. But to be totally sure you could test this, or even better add a unit test.

pedrobaeza replied:
@rven are you going to apply these improvement?

ivantodorovich commented:
Unit price is automatically recalculated whenever you change the quantity. It's very annoying standard behaviour.

Since we are adding a button to recalculate prices, shouldn't we try to avoid this standard behaviour?

![wRCfSpTLo3](https://user-images.githubusercontent.com/1914185/55550851-a35a3200-56d9-11e9-8849-a97e90a21b07.gif)

ivantodorovich reviewed (approved):
Tested on runbot

LGTM 👍

pedrobaeza commented:
> Unit price is automatically recalculated whenever you change the quantity. It's very annoying standard behaviour.

It makes sense when you have escalated prices. Any way, this should be a separated module IMO, but you can combine both.

ivantodorovich commented:
> > Unit price is automatically recalculated whenever you change the quantity. It's very annoying standard behaviour.
> 
> It makes sense when you have escalated prices. Any way, this should be a separated module IMO, but you can combine both.

What about a setting in the Sale's Configuration view, under this very same module?
To be honest, I want to avoid thinking of a name for this new module  😅

pedrobaeza commented:
OK, that can be an option, but it can be added later as improvement with other PR.

ivantodorovich commented:
> OK, that can be an option, but it can be added later as improvement with other PR.

Yes of course

sergio-teruel committed:
[11.0][FIX] sale_order_price_recalculation: set discount from pricelist with discount_policy without_discount

pedrobaeza commented:
Please take into account #833 here

rven commented:
> Please take into account #833 here

Done!

mohamedhagag commented:
Hi all,
can we merge this ?
